README.md

A simple way to return well formatted json responses in a Laravel application.

[![Latest Version on Packagist](https://img.shields.io/packagist/v/tequilarapido/api-response.svg?style=flat-square)](https://packagist.org/packages/tequilarapido/api-response)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/tequilarapido/api-response/master.svg?style=flat-square)](https://travis-ci.org/tequilarapido/api-response)
[![StyleCI](https://styleci.io/repos/70261592/shield)](https://styleci.io/repos/70261592)
[![Quality Score](https://img.shields.io/scrutinizer/g/tequilarapido/api-response.svg?style=flat-square)](https://scrutinizer-ci.com/g/tequilarapido/api-response)
[![Code Coverage](https://img.shields.io/scrutinizer/coverage/g/tequilarapido/api-response/master.svg?style=flat-square)](https://scrutinizer-ci.com/g/tequilarapido/api-response/?branch=master)

<p align="center">
    <img src="https://s9.postimg.org/fcsm1oe8v/illustration.jpg" alt="Laravel Translation Sheet">
</p>

## Installation

You can install the package using composer

``` bash
$ composer require tequilarapido/api-response
```

Add the service provider

``` php
Tequilarapido\ApiResponse\ApiResponseServiceProvider::class 
```


## Usage

This package comes with a helper function `api_response()`, as sugar syntax to using `app(Tequilarapido\ApiResponse\ApiResponse::class)`

 
### Return a response from an item (Array / Associative array / Object/)
 
``` php
return api_response()->item(['result`' => 'success'])
```

Result :   
```
    {
        "data":
        {   
            "result":
            "success"
        }
    }
```

### Return a response from a collection
 
``` php
return api_response()->item(collect(['Apple', 'Orange', 'Kiwi']))
```

Result :   
```   
    {
        "data":
            {   
                "Apple",
                "Orange",
                "Kiwi",
            }
    }
```

### Return a response from a paginated collection

``` php
return api_response()->item(collect(['Apple', 'Orange', 'Kiwi']))
```

Result :   
```
    {
        "data": [1, 2, 3],
        "meta": {
            "pagination": {
                "count": 3,
                "current_page": 1,
                "links": {
                    "next": "/?page=2"
                },
                "per_page": 3,
                "total": 15,
                "total_pages": 5
            }
         }
    }
```

### Transformers 

You can use transformers along with `item()`, `collection()`, and `paginatedCollection` methods to transform results
before returning the responses.

To learn more about the concept behind transformers, please read the [League\Fractal documentation](http://fractal.thephpleague.com/transformers/). 

Put in a simple way, a Transformer class 
    - must extends `League\Fractal\TransformerAbstract` 
    - and contains a method `transform()` that take the raw value/object as an argument and transform it the way we want.
    
Let's assume that we need to return an api response containing a collection of books retreived from database, and we want to format/enhance/restrict results.
  
``` php
class BookTransformer extends TransformerAbstract
{
    public function transform($book)
    {
        return [
            'id' => (int)$book->id,
            'title' => strtoupper($book->title),
            'price' => '$' . $book->price,
            'published_at' => $book->published_at->format('d/m/Y'),
            'url' => 'https://store.books.com/books/' . $book->id
        ];
    }
}
``` 
  
``` php
Route::get('/books', function () {

    // ie. Book::all()
   $books = collect([
       (object)['title' => 'An awesome book', 'id' => '1', 'price' => 10, 'published_at' => Carbon::createFromFormat('Y-m-d', '2016-10-01')],
       (object)['title' => 'A second awesome book', 'id' => '2', 'price' => 100, 'published_at' => Carbon::createFromFormat('Y-m-d', '2016-10-02')],
   ]);

   return api_response()->collection($books, new BookTransformer);
});
```

Result: 
``` 
{
    "data": [
        {
            "id": 1,
            "price": "$10",
            "published_at": "01/10/2016",
            "title": "AN AWESOME BOOK",
            "url": "https://store.books.com/books/1"
        },
        {
            "id": 2,
            "price": "$100",
            "published_at": "02/10/2016",
            "title": "A SECOND AWESOME BOOK",
            "url": "https://store.books.com/books/2"
        }
    ]
}
``` 

### Attaching cookies

For `item()`, `collection()`, and `paginatedCollection` methods the returned result is a built `Illuminate\Http\Response` object.

To be able to attach cookies you need to instruct `api_response()` methods to not build the response by setting the $built argument to false, attach the cookie
and then build the response.

``` php
return api_response()
    ->item(['result`' => 'success'], null, null, false)
    ->withCookie(new \Symfony\Component\HttpFoundation\Cookie('name', 'value'))
    ->build();
```

### Attaching headers

For `item()`, `collection()`, and `paginatedCollection` methods the returned result is a built `Illuminate\Http\Response` object.

To be able to attach headers you need to instruct `api_response()` methods to not build the response by setting the $built argument to false, attach the header
and then build the response.

``` php
return api_response()
    ->item(['result`' => 'success'], null, null, false)
    ->withHeader('X-CUSTOM', 'customvalue')
    ->build();
```

###  Other useful methods 

| Method      | Usage        | 
|-------------|--------------|
| `api_response()->noContent()` | Return an empty response with 204 No Content header.|
| `api_response()->errorNotFound()` | Return a 404 Not found.|
| `api_response()->errorBadRequest()` | Return a 404 Bad Request.|
| `api_response()->errorForbidden()` | Return a 403 Forbidden.|
| `api_response()->errorUnAUthorized()` | Return a 401 Unauthorized.|
| `api_response()->errorInternal()` | Return a 500 Internal Error.|
| `api_response()->error()` | Return a more customizable error (Accept MessageBag instance, staus code ...)|


## Changelog
Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Security

If you discover any security related issues, please email :author_email instead of using the issue tracker.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Credits

- [Nassif Bourguig](https://github.com/nbourguig)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.1.6"></a>
## [1.1.6](https://github.com/tequilarapido/api-response/compare/v1.1.5...v1.1.6) (2020-03-03)



<a name="1.1.5"></a>
## [1.1.5](https://github.com/tequilarapido/api-response/compare/v1.1.4...v1.1.5) (2019-10-10)


### Features

* support Laravel 6.* versions ([c663022](https://github.com/tequilarapido/api-response/commit/c663022))



<a name="1.1.4"></a>
## [1.1.4](https://github.com/tequilarapido/api-response/compare/v1.1.3...v1.1.4) (2019-10-10)


### Features

* support Laravel 6.* versions ([79f79cf](https://github.com/tequilarapido/api-response/commit/79f79cf))



<a name="1.1.3"></a>
## [1.1.3](https://github.com/tequilarapido/api-response/compare/v1.1.2...v1.1.3) (2019-09-04)


### Bug Fixes

* remove hard coded version in composer.json ([3cc49b5](https://github.com/tequilarapido/api-response/commit/3cc49b5))



<a name="1.1.2"></a>
## [1.1.2](https://github.com/tequilarapido/api-response/compare/v1.1.1...v1.1.2) (2019-09-04)



<a name="1.1.1"></a>
## [1.1.1](https://github.com/tequilarapido/api-response/compare/v1.1.0...v1.1.1) (2019-09-03)


### Features

* Add laravel 6.0 support ([0203b29](https://github.com/tequilarapido/api-response/commit/0203b29))



# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

# 1.1.0 (2019-03-04)


### Features

* Add laravel 5.8 support ([b418cb2](https://github.com/tequilarapido/api-response/commit/b418cb2))
  BC: phpunit 8.0 tests.


tests/Kernel.php
<?php

namespace Tequilarapido\ApiResponse\Test;

use Exception;

class Kernel extends \Illuminate\Foundation\Console\Kernel
{
    /**
     * The bootstrap classes for the application.
     */
    protected $bootstrappers = [];

    /**
     * The Artisan commands provided by your application.
     *
     * @var array
     */
    protected $commands = [];

    /**
     * Report the exception to the exception handler.
     *
     * @param \Exception $e
     *
     * @throws \Exception
     */
    protected function reportException(Exception $e)
    {
        throw $e;
    }

    public function getArtisan()
    {
        return $this->app['artisan'];
    }
}


tests/TestCase.php
<?php

namespace Tequilarapido\ApiResponse\Test;

use Tequilarapido\ApiResponse\ApiResponseServiceProvider;

class TestCase extends \Orchestra\Testbench\TestCase
{
    /** @var TestHelper */
    protected $helper;

    protected function getPackageProviders($app)
    {
        return [ApiResponseServiceProvider::class];
    }

    public function setUp() : void
    {
        parent::setUp();

        $this->helper = new TestHelper($this->app);
    }

    public function tearDown() : void
    {
        parent::tearDown();

        $this->helper = null;
    }
}


tests/TestHelper.php
<?php

namespace Tequilarapido\ApiResponse\Test;

class TestHelper
{
    /** @var \Illuminate\Foundation\Application */
    protected $app;

    /**
     * TestHelper constructor.
     *
     * @param $app
     */
    public function __construct($app)
    {
        $this->app = $app;
    }

    /**
     * Call protected/private method of a class.
     *
     * @param object &$object    Instantiated object that we will run method on.
     * @param string $methodName Method name to call
     * @param array  $parameters Array of parameters to pass into method.
     *
     * @return mixed Method return.
     */
    public function invokeMethod(&$object, $methodName, array $parameters = [])
    {
        $reflection = new \ReflectionClass(get_class($object));
        $method = $reflection->getMethod($methodName);
        $method->setAccessible(true);

        return $method->invokeArgs($object, $parameters);
    }
}


src/FractalAdapter.php
<?php

namespace Tequilarapido\ApiResponse;

use Illuminate\Contracts\Pagination\LengthAwarePaginator;
use Illuminate\Contracts\Support\Arrayable;
use Illuminate\Http\Request;
use League\Fractal\Manager;
use League\Fractal\Resource\Collection;
use League\Fractal\Resource\Item;
use League\Fractal\Serializer\SerializerAbstract;

class FractalAdapter
{
    /** @var \League\Fractal\Manager */
    protected $manager;

    /** @var Request */
    private $request;

    /**
     * FractalAdapter constructor.
     *
     * @param SerializerAbstract $serializer
     * @param Request $request
     */
    public function __construct(SerializerAbstract $serializer, Request $request)
    {
        $this->manager = new Manager();
        $this->manager->setSerializer($serializer);

        $this->request = $request;
    }

    /**
     * Return an item.
     *
     * @param $data
     * @param null $transformer
     * @param null $resourceKey
     * @return array
     */
    public function item($data, $transformer = null, $resourceKey = null)
    {
        $resource = new Item($data, $this->getTransformer($transformer), $resourceKey);

        return $this->manager->createData($resource)->toArray();
    }

    /**
     * Return a collection of items.
     *
     * @param $data
     * @param null $transformer
     * @param null $resourceKey
     * @return array
     */
    public function collection($data, $transformer = null, $resourceKey = null)
    {
        $resource = new Collection($data, $this->getTransformer($transformer), $resourceKey);

        return $this->manager->createData($resource)->toArray();
    }

    /**
     * @param LengthAwarePaginator $paginator
     * @param null $transformer
     * @param null $resourceKey
     * @return array
     */
    public function paginatedCollection(LengthAwarePaginator $paginator, $transformer = null, $resourceKey = null)
    {
        $paginator->appends($this->request->query());

        $resource = new Collection($paginator->items(), $this->getTransformer($transformer), $resourceKey);

        $resource->setPaginator(new PaginatorAdapter($paginator));

        return $this->manager->createData($resource)->toArray();
    }

    /**
     * Return transformer.
     *
     * @param $transformer
     * @return \Closure
     */
    protected function getTransformer($transformer)
    {
        return $transformer ?: function ($data) {
            if ($data instanceof Arrayable) {
                return $data->toArray();
            }

            return (array) $data;
        };
    }
}


src/ResponseBuilder.php
<?php

namespace Tequilarapido\ApiResponse;

use Illuminate\Http\Response as HttpResponse;
use Symfony\Component\HttpFoundation\Cookie;

/**
 * Inspired from dingo/api.
 *
 * Class ResponseBuilder
 */
class ResponseBuilder
{
    /**
     * Response content.
     *
     * @var mixed
     */
    protected $response;

    /**
     * The HTTP response headers.
     *
     * @var array
     */
    protected $headers = [];

    /**
     * The HTTP response cookies.
     *
     * @var array
     */
    protected $cookies = [];

    /**
     * The HTTP response status code.
     *
     * @var int
     */
    protected $statusCode = 200;

    /**
     * Create a new response builder instance.
     *
     * @param mixed $response
     */
    public function __construct($response)
    {
        $this->response = $response;
    }

    /**
     * Add a cookie to the response.
     *
     * @param \Symfony\Component\HttpFoundation\Cookie $cookie
     *
     * @return ResponseBuilder
     */
    public function withCookie(Cookie $cookie)
    {
        $this->cookies[] = $cookie;

        return $this;
    }

    /**
     * Add a cookie to the response.
     *
     * @param \Symfony\Component\HttpFoundation\Cookie $cookie
     *
     * @return ResponseBuilder
     */
    public function cookie(Cookie $cookie)
    {
        return $this->withCookie($cookie);
    }

    /**
     * Add a header to the response.
     *
     * @param string $name
     * @param string $value
     *
     * @return ResponseBuilder
     */
    public function withHeader($name, $value)
    {
        $this->headers[$name] = $value;

        return $this;
    }

    /**
     * Add a header to the response.
     *
     * @param string $name
     * @param string $value
     *
     * @return ResponseBuilder
     */
    public function header($name, $value)
    {
        return $this->withHeader($name, $value);
    }

    /**
     * Set the responses status code.
     *
     * @param int $statusCode
     *
     * @return ResponseBuilder
     */
    public function setStatusCode($statusCode)
    {
        $this->statusCode = $statusCode;

        return $this;
    }

    /**
     * Set the response status code.
     *
     * @param int $statusCode
     *
     * @return ResponseBuilder
     */
    public function statusCode($statusCode)
    {
        return $this->setStatusCode($statusCode);
    }

    /**
     * Build the response.
     *
     * @return \Illuminate\Http\Response
     */
    public function build()
    {
        $response = new HttpResponse($this->response, $this->statusCode, $this->headers);
        
        // Force json content type
        $response->header('Content-Type', 'application/json');

        foreach ($this->cookies as $cookie) {
            if ($cookie instanceof Cookie) {
                $response->withCookie($cookie);
            }
        }

        return $response;
    }

    /**
     * Returns cookies.
     *
     * @return array
     */
    public function getCookies()
    {
        return $this->cookies;
    }

    /**
     * Return headers.
     *
     * @return array
     */
    public function getHeaders()
    {
        return $this->headers;
    }

    /**
     * Return status code.
     *
     * @return int
     */
    public function getStatusCode()
    {
        return $this->statusCode;
    }
    
    /**
     * Add meta to response
     *
     * @param $key
     * @param $value
     * @return $this
     */
    public function setMeta($key, $value)
    {
        if (!isset($this->response['meta'])) {
            $this->response['meta'] = [];
        }

        $this->response['meta'][$key] = $value;
        
        return $this;
    }
}


src/PaginatorAdapter.php
<?php

namespace Tequilarapido\ApiResponse;

use Illuminate\Contracts\Pagination\LengthAwarePaginator;
use Illuminate\Contracts\Pagination\Paginator;
use League\Fractal\Pagination\PaginatorInterface;

class PaginatorAdapter implements PaginatorInterface
{
    /**
     * @var LengthAwarePaginator
     */
    private $paginator;

    /**
     * PaginatorAdapter constructor.
     *
     * @param LengthAwarePaginator $paginator
     */
    public function __construct(LengthAwarePaginator $paginator)
    {
        $this->paginator = $paginator;
    }

    /**
     * @return int
     */
    public function getCurrentPage()
    {
        return $this->paginator->currentPage();
    }

    /**
     * @return int
     */
    public function getLastPage()
    {
        return $this->paginator->lastPage();
    }

    /**
     * @return int
     */
    public function getTotal()
    {
        return $this->paginator->total();
    }

    /**
     * @return int
     */
    public function getCount()
    {
        return $this->paginator->count();
    }

    /**
     * @return int
     */
    public function getPerPage()
    {
        return $this->paginator->perPage();
    }

    /**
     * @param int $page
     *
     * @return string
     */
    public function getUrl($page)
    {
        return $this->paginator->url($page);
    }

    /**
     * @return Paginator
     */
    public function getPaginator()
    {
        return $this->paginator;
    }
}


src/ApiResponse.php
<?php

namespace Tequilarapido\ApiResponse;

use Illuminate\Support\MessageBag;

class ApiResponse
{
    /**
     * @var FractalAdapter
     */
    private $fractalAdapter;

    /**
     * Response class.
     *
     * @param FractalAdapter $fractal
     */
    public function __construct(FractalAdapter $fractal)
    {
        $this->fractalAdapter = $fractal;
    }

    /**
     * Returns one item.
     *
     * @param mixed $data
     * @param null  $transformer
     * @param null  $resourceKey
     * @param bool  $build
     *
     * @return array|\Illuminate\Http\Response|ResponseBuilder
     */
    public function item($data, $transformer = null, $resourceKey = null, $build = true)
    {
        return $this->toResponseBuilder(
            $this->fractalAdapter->item($data, $transformer, $resourceKey),
            $build
        );
    }

    /**
     * Returns collection.
     *
     * @param      $data
     * @param null $transformer
     * @param null $resourceKey
     * @param bool $build
     *
     * @return array|\Illuminate\Http\Response|ResponseBuilder
     */
    public function collection($data, $transformer = null, $resourceKey = null, $build = true)
    {
        return $this->toResponseBuilder(
            $this->fractalAdapter->collection($data, $transformer, $resourceKey),
            $build
        );
    }

    /**
     * Returns paginated collection.
     *
     * @param      $paginator
     * @param null $transformer
     * @param null $resourceKey
     * @param bool $build
     *
     * @return array|\Illuminate\Http\Response|ResponseBuilder
     */
    public function paginatedCollection($paginator, $transformer = null, $resourceKey = null, $build = true)
    {
        return $this->toResponseBuilder(
            $this->fractalAdapter->paginatedCollection($paginator, $transformer, $resourceKey),
            $build
        );
    }

    /**
     * Respond with a no content response.
     *
     * @param bool $build
     *
     * @return ResponseBuilder|\Illuminate\Http\Response
     */
    public function noContent($build = true)
    {
        $response = new ResponseBuilder(null);
        $response->setStatusCode(204);

        return $build ? $response->build() : $response;
    }

    /**
     * Return a 404 not found error.
     *
     * @param string|array $message
     * @param bool         $build
     *
     * @return \Illuminate\Http\Response
     */
    public function errorNotFound($message = 'Not Found', $build = true)
    {
        return $this->error($message, 404, null, $build);
    }

    /**
     * Return a 400 bad request error.
     *
     * @param string|array $message
     * @param bool         $build
     *
     * @return \Illuminate\Http\Response
     */
    public function errorBadRequest($message = 'Bad Request', $build = true)
    {
        return $this->error($message, 400, null, $build);
    }

    /**
     * Return a 401 unauthorized error.
     *
     * @param string|array $message
     * @param bool         $build
     *
     * @return \Illuminate\Http\Response
     */
    public function errorUnauthorized($message = 'Unauthorized', $build = true)
    {
        return $this->error($message, 401, null, $build);
    }

    /**
     * Return a 401 unauthorized error.
     *
     * @param string|array $message
     * @param bool         $build
     *
     * @return \Illuminate\Http\Response
     */
    public function errorForbidden($message = 'Forbidden', $build = true)
    {
        return $this->error($message, 403, null, $build);
    }

    /**
     * Return a 500 internal server error.
     *
     * @param string|array $message
     * @param bool         $build
     *
     * @return \Illuminate\Http\Response
     */
    public function errorInternal($message = 'Internal Error', $build = true)
    {
        return $this->error($message, 500, null, $build);
    }

    /**
     * Return an error response.
     *
     * @param                  $messages
     * @param int              $statusCode
     * @param mixed|MessageBag $errors
     * @param bool             $build
     *
     * @return \Illuminate\Http\Response
     */
    public function error($messages, $statusCode = 500, $errors = null, $build = true)
    {
        $error = [
            'error' => true,
            'code' => $statusCode,
            'message' => $messages,
        ];

        if (! is_null($errors)) {
            $error = array_merge($error, ['errors' => $errors]);
        }

        $response = (new ResponseBuilder($error))->setStatusCode($statusCode);

        return $build ? $response->build() : $response;
    }

    /**
     * Wrap into ResponseBuilder, and build if requested.
     * If not will return ResponseBuilder object, to be able to add headers for instance
     * before sending back response using `build()` method.
     *
     * @param $data
     * @param bool $build
     *
     * @return \Illuminate\Http\Response|ResponseBuilder
     */
    protected function toResponseBuilder($data, $build)
    {
        return $build
            ? (new ResponseBuilder($data))->build()
            : new ResponseBuilder($data);
    }
}
